### PR TITLE
Fixes adding Private Media from same S3 Disk

### LINKED ISSUE #2203: Media is not uploaded with proper custom headers on S3
When uploading a media that has custom headers for S3, if the remote file which is used to upload media is on the same disk then the file is copied directly. During this process, the custom headers expected in the new file are not inserted.

This is specifically happening in v8 of Media Library as the copy logic was not there till v7.
Check copyToMediaLibraryFromRemote() function in file src/MediaCollections/Filesystem.php line 52 of v8.

E.g.
1.Temp file is uploaded with private ACL.
2. Now when that file is used to upload to the new public location public ACL is provided through medialibrary.php config file.
3. Now maybe to improve performance, if both the files are on the same disk, they are just copied over and not re-retrieved and uploaded.
4. As the file is just copied over the extra custom headers are never applied.

### PR DESCRIPTION
**Pull request by ebbbang**
Fixes #2203

### CODE REVIEW

File: src/MediaCollections/Filesystem.php [hunk 1]
```diff
@@ -77,6 +77,11 @@ public function copyToMediaLibraryFromRemote(RemoteFile $file, Media $media, ?st
         );
     }
 
+    protected function shouldCopyFileOnDisk(RemoteFile $file, Media $media, string $diskDriverName): bool
+    {
+        return $file->getDisk() === $media->disk && ($diskDriverName !== 's3' || count($media->getCustomHeaders()) === 0);
```
Comment (freekmurze) [inline src/MediaCollections/Filesystem.php:82]: This kind of conditional is unreadable to me. Could you split this up and use early returns?

**Review by freekmurze — CHANGES_REQUESTED** [review]

File: src/MediaCollections/Filesystem.php (hunk 1 above)
Comment (ebbbang) [reply src/MediaCollections/Filesystem.php]: Done, please let me know if that helps. If not I'll be more than happy to improve it further.

For your reference, the conditions I want to meet are:
1. It only affects if the disks are the same.
2. It only affects S3 as far as I know.
3. And only when custom headers are involved which were not present on the source file, to begin with.

File: src/MediaCollections/Filesystem.php [hunk 2]
```diff
@@ -77,6 +77,23 @@ public function copyToMediaLibraryFromRemote(RemoteFile $file, Media $media, ?st
         );
     }
 
+    protected function shouldCopyFileOnDisk(RemoteFile $file, Media $media, string $diskDriverName): bool
+    {
+        if ($file->getDisk() !== $media->disk) {
+            return false;
+        }
+
+        if ($diskDriverName !== 's3') {
```
Comment (freekmurze) [inline src/MediaCollections/Filesystem.php:86]: Checking "s3" here feels a bit strange. Wouldn't it be better to check for `=== 'local'` here?

File: src/MediaCollections/Filesystem.php (hunk 1 above)
Comment (freekmurze) [reply src/MediaCollections/Filesystem.php]: Thanks!

File: src/MediaCollections/Filesystem.php (hunk 2 above)
Comment (ebbbang) [reply src/MediaCollections/Filesystem.php:86]: Yes, we could use ```=== 'local'``` here, but I did not know how that would affect other filesystem drivers performance-wise.

I assumed that copy would be most efficient for most drivers and even for 's3' if no custom headers are provided.

If you want or if I'm missing something then let me know and I can change it to ```=== 'local'```.

Comment (ebbbang) [reply src/MediaCollections/Filesystem.php:86]: @freekmurze Please do let me know what you think?

Comment (freekmurze) [reply src/MediaCollections/Filesystem.php:86]: Let's go with `=== 'local'` 👍

Comment (ebbbang) [reply src/MediaCollections/Filesystem.php]: Done. I also squashed the commits in the process to keep it clean. Please let me know if more work is needed or can we merge it now?

**Comment by freekmurze** [comment]
Thanks!

**Pull request merged**